### PR TITLE
NestedFolderPicker: Seperate state from Browse Dashboards

### PR DESCRIPTION
**Pull request by joshhunt**
Refactors the nested folder picker to not use the Browse Dashboards redux state. Instead, we create a new RTKQ query (listFolders) and manually take care of subscriptions and selecting the data out of the redux store. Because the requests are properly cached by all their arguments, we don't risk 'cache poisoning' by loading different results into the browse dashboards cache.

Fixes an issue where fully loading a folder in the folder picker could prevent dashboards from being listed in the Browse Dashboards list.

### CODE REVIEW

File: public/app/features/browse-dashboards/api/browseDashboardsAPI.ts [hunk 1]
```diff
@@ -69,11 +69,41 @@ function createBackendSrvBaseQuery({ baseURL }: { baseURL: string }): BaseQueryF
...
+}
+
 export const browseDashboardsAPI = createApi({
   tagTypes: ['getFolder'],
   reducerPath: 'browseDashboardsAPI',
   baseQuery: createBackendSrvBaseQuery({ baseURL: '/api' }),
   endpoints: (builder) => ({
+    listFolders: builder.query<FolderDTO[], ListFolderArgs>({
+      // TODO: provides tags
+      query: ({ page, parentUid, limit }) => ({ url: '/folders', params: { page, parentUid, limit } }),
```
Comment (joshhunt) [inline public/app/features/browse-dashboards/api/browseDashboardsAPI.ts:86]: Turns out if we just remember the arguments we used for the requests, we can just select them each individually from the store, and don't need to merge them together.

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx [hunk 2]
```diff
@@ -385,3 +396,153 @@ const getStyles = (theme: GrafanaTheme2) => {
...
+  (state, parentUid, page, limit) => {
+    return browseDashboardsAPI.endpoints.listFolders.select({ parentUid, page, limit })(state);
+  }
+);
+
+function useFolderList() {
+  const requestedArgs = useRef<ListFolderArgs[] | null>(null);
+  if (requestedArgs.current === null) {
+    requestedArgs.current = [];
+  }
```
Comment (joshhunt) [inline public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:430]: Store references to the data we've requested

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx [hunk 3]
```diff
@@ -385,3 +396,153 @@ const getStyles = (theme: GrafanaTheme2) => {
...
+      let page = 0;
+      if (!pageSet) {
+        page = 0;
+      } else {
+        page = pageSet.length;
+      }
+
+      const args = { parentUid, page, limit: PAGE_SIZE };
+
+      requestedArgs.current?.push(args);
```
Comment (joshhunt) [inline public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:530]: Step 1: When making a request, remember the arguments to select them later

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx [hunk 4]
```diff
@@ -385,3 +396,153 @@ const getStyles = (theme: GrafanaTheme2) => {
...
+
+      const promise = dispatch(browseDashboardsAPI.endpoints.listFolders.initiate(args));
```
Comment (joshhunt) [inline public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:532]: Step 1: When making a request, remember the arguments to select them later

Comment (joshhunt) [inline public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:532]: Step 1: When making a request, remember the arguments to select them later

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx [hunk 5]
```diff
@@ -385,3 +396,153 @@ const getStyles = (theme: GrafanaTheme2) => {
...
+  const state = useSelector((rxState) => {
+    const requests = requestedArgs.current ?? [];
+
+    const pages = requests
+      .map((args) => {
+        return listFolderSelector(rxState, args.parentUid, args.page, args.limit);
+      })
+      .filter((page, index, all) => {
+        return all.findIndex((otherPage) => otherPage.requestId === page.requestId) === index;
+      });
```
Comment (joshhunt) [inline public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:442]: Step 2: Then, when the store updates, loop over the requests and select the data from RTKQ to use

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx [hunk 6]
```diff
@@ -385,3 +396,153 @@ const getStyles = (theme: GrafanaTheme2) => {
...
+  });
+
+  console.log('state', state);
+
+  const rootCollection = useMemo(() => {
+    const rrr = state.rootPages;
+
+    let flatTree: Array<DashboardsTreeItem<DashboardViewItemWithUIItems>> = rrr.flatMap((page) => {
+      return (page.data ?? []).map((item) => {
+        const ddddd: DashboardsTreeItem<DashboardViewItemWithUIItems> = {
```
Comment (joshhunt) [inline public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:474]: Step 3, 4, ... 100: Then construct the flat tree from that state we selected out.

I think for I will re-create the createFlatTree function just for this component, but I've thought a little about how we could generalise it through selector functions.

File: .betterer.results.json
Comment (joshhunt) [inline .betterer.results.json:1]: not really sure what's going on here... ill be sure to rebase before i put this up for review properly

File: public/app/core/components/NestedFolderPicker/api.ts [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+      return flatList;
+    }
+
+    const rootFlatTree = createFlatList(undefined, state.rootPages, 1);
+    rootFlatTree.unshift(ROOT_FOLDER_ITEM);
+
+    return rootFlatTree;
+  }, [state, isBrowsing, openFolders]);
+
+  return [treeList, state.isLoading, requestNextPage] as const;
```
Comment (Clarity-89) [inline public/app/core/components/NestedFolderPicker/api.ts:195]: I think using object return would be better here so you don't need to remember the order.

**Review by Clarity-89 — APPROVED** [review]
Tested this earlier today and didn't find any issues 🚀

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx [hunk 8]
```diff
@@ -66,35 +54,41 @@ async function getSearchResults(searchQuery: string) {
 export function NestedFolderPicker({
   value,
   invalid,
-  showRootFolder = true,
+  showRootFolder = true, // JOSH TODO: Restore this
   clearable = false,
-  excludeUIDs,
+  excludeUIDs, // JOSH TODO: Restore this (or not, if it's not needed)
```
Comment (ashharrison90) [inline public/app/core/components/NestedFolderPicker/NestedFolderPicker.tsx:59]: can these `// TODO` comments go?

File: public/app/core/components/NestedFolderPicker/api.ts
Comment (ashharrison90) [inline public/app/core/components/NestedFolderPicker/api.ts:1]: feels a little weird that we have a file called api containing a mixture of selectors and hooks 🤔 on the one hand, i think we could split it into `selectors.ts` and `hooks.ts`. on the other, it's kinda useful to see all this logic together.... wdyt?

File: public/app/core/components/NestedFolderPicker/api.ts [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    (
+      parentUid: ListFolderQueryArgs['parentUid'],
+      page: ListFolderQueryArgs['page'],
+      limit: ListFolderQueryArgs['limit']
+    ) => limit,
+  ],
+  (parentUid, page, limit) => {
+    return browseDashboardsAPI.endpoints.listFolders.select({ parentUid, page, limit });
+  }
+);
```
Comment (ashharrison90) [inline public/app/core/components/NestedFolderPicker/api.ts:39]: do we need this? 🤔 feels like we can just `return browseDashboardsAPI.endpoints.listFolders.select({ parentUid, page, limit });` on line 48 directly

**Review by ashharrison90 — COMMENTED** [review]
_great_ job on this 🚀 

everything works as expected as far as i can tell, just left a couple of nits to consider 🙌

File: public/app/core/components/NestedFolderPicker/api.ts
Comment (joshhunt) [reply public/app/core/components/NestedFolderPicker/api.ts:1]: I know what you mean. We already have a `hooks.ts` that contains the keyboard interaction hooks which is already a bit meaty, so I didn't want to add all of this to that as well.

I think I would prefer to keep them seperate. Could rename hooks.ts to `interactions.ts` and/or this one to... `query.ts`?

File: public/app/core/components/NestedFolderPicker/api.ts (hunk 9 above)
Comment (joshhunt) [reply public/app/core/components/NestedFolderPicker/api.ts:39]: ahh hmm, good catch. I'll be honest, I have no idea what's going on here - getting the selector right was actually probably the hardest part of the PR 😅

I'm happy to give this a try, but if it doesn't immediately work I'll ask you for working code if you want it changed 💜

File: public/app/core/components/NestedFolderPicker/NestedFolderPicker.test.tsx [hunk 10]
```diff
@@ -127,6 +150,40 @@ describe('NestedFolderPicker', () => {
...
+  it('hides folders specififed by UID', async () => {
+    render(<NestedFolderPicker excludeUIDs={[folderA.item.uid]} onChange={mockOnChange} />);
+
+    // Open the picker and wait for children to load
+    const button = await screen.findByRole('button', { name: 'Select folder' });
+    await userEvent.click(button);
+    await screen.findByLabelText(folderB.item.title);
+
+    expect(screen.queryByLabelText(folderA.item.title)).not.toBeInTheDocument();
+  });
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/core/components/NestedFolderPicker/NestedFolderPicker.test.tsx:185]: 😊

File: public/app/core/components/NestedFolderPicker/api.ts
Comment (joshhunt) [reply public/app/core/components/NestedFolderPicker/api.ts:1]: renamed the files to the hooks in them - `useTreeInteractions` and `useFoldersQuery`

Comment (ashharrison90) [reply public/app/core/components/NestedFolderPicker/api.ts:1]: i was initially just commenting on the fact that the selectors are kinda grouped in with the hook, but we do the same in browse dashboards so after pondering a bit more i think it's fine 😄 didn't even notice the keyboard hooks.ts file already existing, but i like the rename you've done 👍

File: public/app/core/components/NestedFolderPicker/api.ts (hunk 9 above)
Comment (joshhunt) [reply public/app/core/components/NestedFolderPicker/api.ts]: This worked. Thanks for the fresh eyes Ash!! 👏

**Review by ashharrison90 — APPROVED** [review]
lgtm 🚀

**Pull request merged**